### PR TITLE
Allow upstream timeout to return 204 when indicated by header

### PR DESCRIPTION
**Pull request by goaway**

### CODE REVIEW

File: source/common/http/headers.h [hunk 1]
```diff
@@ -27,6 +27,8 @@ class Headers {
   const LowerCaseString EnvoyUpstreamAltStatName{"x-envoy-upstream-alt-stat-name"};
   const LowerCaseString EnvoyUpstreamCanary{"x-envoy-upstream-canary"};
   const LowerCaseString EnvoyUpstreamRequestTimeoutMs{"x-envoy-upstream-rq-timeout-ms"};
+  const LowerCaseString EnvoyUpstreamRequestTimeoutAltResponse{
+      "x-envoy-upstream_request-timeout-alt-response"};
```
Comment (mattklein123) [inline source/common/http/headers.h]: x-envoy-upstream-request-timeout-alt-response ?

File: source/common/router/router.h [hunk 2]
```diff
@@ -206,6 +206,7 @@ class Filter : Logger::Loggable<Logger::Id::router>, public Http::StreamDecoderF
   const VirtualCluster* request_vcluster_;
   Event::TimerPtr response_timeout_;
   FilterUtility::TimeoutData timeout_;
+  Http::Code timeout_response_code_;
```
Comment (mattklein123) [inline source/common/router/router.h]: just initialize to Http::Code::GatewayTimeout here, and avoid else case in other file.

File: source/common/router/router.cc [hunk 3]
```diff
@@ -191,6 +191,17 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
   }
 
   timeout_ = FilterUtility::finalTimeout(*route_entry_, headers);
+
+  // If this header is set with any value, use an alternate response code on timeout
```
Comment (mattklein123) [inline source/common/router/router.cc]: do we want to allow user to specify response code in header? Seems more flexible?

File: source/common/router/router.h (hunk 2 above)
Comment (goaway) [inline source/common/router/router.h]: sure

File: source/common/router/router.cc (hunk 3 above)
Comment (goaway) [inline source/common/router/router.cc]: I guess we can do that - any downside to allowing an arbitrary response code here?

File: source/common/http/headers.h (hunk 1 above)
Comment (goaway) [inline source/common/http/headers.h]: oops

File: include/envoy/http/header_map.h [hunk 4]
```diff
@@ -196,6 +196,7 @@ class HeaderEntry {
   HEADER_FUNC(EnvoyUpstreamHealthCheckedCluster)                                                   \
   HEADER_FUNC(EnvoyUpstreamRequestPerTryTimeoutMs)                                                 \
   HEADER_FUNC(EnvoyUpstreamRequestTimeoutMs)                                                       \
+  HEADER_FUNC(EnvoyUpstreamRequestTimeoutAltResponse)                                              \
```
Comment (RomanDzhabarov) [inline include/envoy/http/header_map.h]: alpha order

File: source/common/http/headers.h [hunk 5]
```diff
@@ -27,6 +27,8 @@ class Headers {
   const LowerCaseString EnvoyUpstreamAltStatName{"x-envoy-upstream-alt-stat-name"};
   const LowerCaseString EnvoyUpstreamCanary{"x-envoy-upstream-canary"};
   const LowerCaseString EnvoyUpstreamRequestTimeoutMs{"x-envoy-upstream-rq-timeout-ms"};
+  const LowerCaseString EnvoyUpstreamRequestTimeoutAltResponse{
```
Comment (RomanDzhabarov) [inline source/common/http/headers.h]: ditto

File: source/common/router/router.cc (hunk 3 above)
Comment (mattklein123) [inline source/common/router/router.cc]: I don't see any downside. The user has to configure it, and we only allow this feature on internal calls.

Comment (goaway) [inline source/common/router/router.cc]: Closing the loop on this - due to format and content restrictions implied by some HTTP status codes, decided to stick with 204 only.

File: docs/configuration/http_filters/router_filter.rst [hunk 6]
```diff
@@ -131,6 +131,15 @@ The output tree is documented :ref:`here <config_cluster_manager_cluster_stats_d
 
 .. _config_http_filters_router_x-envoy-upstream-rq-timeout-ms:
```
Comment (ccaraman) [inline docs/configuration/http_filters/router_filter.rst]: this needs to be updated or deleted to match the new header you added.

Comment (goaway) [inline docs/configuration/http_filters/router_filter.rst]: thanks :)

**Pull request merged**